vowpalwabbit/model_utils.h

```diff
@@ -97,6 +99,64 @@ size_t write_model_field(io_buf& io, const T& var, const std::string& name_or_re
   return details::check_length_matches(io.bin_write_fixed(data, len), len);
 }
 
+size_t read_model_field(io_buf&, std::string&);
+size_t write_model_field(io_buf&, const std::string&, const std::string&, bool);
+template <typename T>
+size_t read_model_field(io_buf&, std::set<T>&);
+template <typename T>
+size_t write_model_field(io_buf&, const std::set<T>&, const std::string&, bool);
+template <typename T>
+size_t read_model_field(io_buf&, std::vector<T>&);
+template <typename T>
+size_t write_model_field(io_buf&, const std::vector<T>&, const std::string&, bool);
+template <typename T>
+size_t read_model_field(io_buf&, v_array<T>&);
+template <typename T>
+size_t write_model_field(io_buf&, const v_array<T>&, const std::string&, bool);
+template <typename F, typename S>
+size_t read_model_field(io_buf&, std::pair<F, S>&);
+template <typename F, typename S>
+size_t write_model_field(io_buf&, const std::pair<F, S>&, const std::string&, bool);
+template <typename T>
+size_t read_model_field(io_buf&, std::priority_queue<T>&);
+template <typename T>
+size_t write_model_field(io_buf&, const std::priority_queue<T>&, const std::string&, bool);
+template <typename K, typename V>
+size_t read_model_field(io_buf&, std::map<K, V>&);
+template <typename K, typename V>
+size_t write_model_field(io_buf&, const std::map<K, V>&, const std::string&, bool);
+
+inline size_t read_model_field(io_buf& io, std::string& str)
+{
+  size_t bytes = 0;
+  uint64_t str_size;
+  bytes += read_model_field(io, str_size);
+  std::ostringstream os;
+  for (uint64_t i = 0; i < str_size; ++i)
+  {
+    char c;
+    bytes += read_model_field(io, c);
+    os << c;
```

Comment:
Rather than reading one by one you could use the c-style string constructor

Reply:
Do you mean making another overload to read_model_field for c-style strings?

Reply:
It's pretty inefficient to serialize/deserialize a string one character at a time. `bin_read_fixed` or something like that seems appropriate here. Since you know the size already you can just memcpy the contents

Reply:
Makes sense, updated